File: modules/microsoft-intune/pkgs/fake-os-release.c
/*
 * LD_PRELOAD library to intercept /etc/os-release reads
 * Returns fake Ubuntu os-release content for Intune compatibility
 */
#define _GNU_SOURCE
#include <dlfcn.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <stdarg.h>
#include <sys/types.h>
#include <sys/stat.h>

static const char FAKE_OS_RELEASE[] =
    "PRETTY_NAME=\"Ubuntu 22.04.5 LTS\"\n"
    "NAME=\"Ubuntu\"\n"
    "VERSION_ID=\"22.04\"\n"
    "VERSION=\"22.04.5 LTS (Jammy Jellyfish)\"\n"
    "VERSION_CODENAME=jammy\n"
    "ID=ubuntu\n";

typedef int (*open_func_t)(const char *, int, ...);

static int create_fake_fd(void) {
    int fd = memfd_create("fake-os-release", MFD_CLOEXEC);
    if (fd < 0) return -1;

    size_t len = sizeof(FAKE_OS_RELEASE) - 1;
    if (write(fd, FAKE_OS_RELEASE, len) != (ssize_t)len) {
        close(fd);
        return -1;
    }
    lseek(fd, 0, SEEK_SET);
    return fd;
}

static int is_os_release(const char *pathname) {
    if (!pathname) return 0;
    return (strcmp(pathname, "/etc/os-release") == 0 ||
            strcmp(pathname, "/usr/lib/os-release") == 0);
}

int open(const char *pathname, int flags, ...) {
    if (is_os_release(pathname)) {
        return create_fake_fd();
    }

    open_func_t real_open = (open_func_t)dlsym(RTLD_NEXT, "open");
    if (!real_open) return -1;

    if (flags & O_CREAT) {
        va_list args;
        va_start(args, flags);
        mode_t mode = va_arg(args, mode_t);
        va_end(args);
        return real_open(pathname, flags, mode);
    }
    return real_open(pathname, flags);
}

int open64(const char *pathname, int flags, ...) {
    if (is_os_release(pathname)) {
        return create_fake_fd();
    }

    open_func_t real_open64 = (open_func_t)dlsym(RTLD_NEXT, "open64");
    if (!real_open64) return -1;

    if (flags & O_CREAT) {
        va_list args;
        va_start(args, flags);
        mode_t mode = va_arg(args, mode_t);
        va_end(args);
        return real_open64(pathname, flags, mode);
    }
    return real_open64(pathname, flags);
}

File: nixos/modules/microsoft-intune/pkgs/fake-os-release.zig
const std = @import("std");
const os = std.os;
const linux = std.os.linux;

const FAKE_OS_RELEASE =
    \\PRETTY_NAME="Ubuntu 22.04.5 LTS"
    \\NAME="Ubuntu"
    \\VERSION_ID="22.04"
    \\VERSION="22.04.5 LTS (Jammy Jellyfish)"
    \\VERSION_CODENAME=jammy
    \\ID=ubuntu
    \\
;

fn createFakeFd() c_int {
    // Use memfd_create syscall directly
    const fd = linux.memfd_create("fake-os-release", linux.MFD.CLOEXEC);
    if (@as(isize, @bitCast(fd)) < 0) return -1;

    const fd_i32: i32 = @intCast(fd);
    const written = linux.write(fd_i32, FAKE_OS_RELEASE, FAKE_OS_RELEASE.len);
    if (written != FAKE_OS_RELEASE.len) {
        _ = linux.close(fd_i32);
        return -1;
    }
    _ = linux.lseek(fd_i32, 0, linux.SEEK.SET);
    return fd_i32;
}

fn isOsRelease(pathname: ?[*:0]const u8) bool {
    if (pathname == null) return false;
    const path = std.mem.span(pathname.?);
    return std.mem.eql(u8, path, "/etc/os-release") or
        std.mem.eql(u8, path, "/usr/lib/os-release");
}

const OpenFn = *const fn ([*:0]const u8, c_int, linux.mode_t) callconv(std.builtin.CallingConvention.c) c_int;

var real_open: ?OpenFn = null;
var real_open64: ?OpenFn = null;

fn getNextOpen(comptime name: [*:0]const u8) ?OpenFn {
    const RTLD_NEXT: *anyopaque = @ptrFromInt(@as(usize, @bitCast(@as(isize, -1))));
    const sym = std.c.dlsym(RTLD_NEXT, name);
    if (sym) |s| {
        return @ptrCast(s);
    }
    return null;
}

export fn open(pathname: ?[*:0]const u8, flags: c_int, mode: linux.mode_t) callconv(std.builtin.CallingConvention.c) c_int {
    if (isOsRelease(pathname)) {
        return createFakeFd();
    }

    if (real_open == null) {
        real_open = getNextOpen("open");
    }
    if (real_open) |func| {
        return func(pathname.?, flags, mode);
    }
    return -1;
}

export fn open64(pathname: ?[*:0]const u8, flags: c_int, mode: linux.mode_t) callconv(std.builtin.CallingConvention.c) c_int {
    if (isOsRelease(pathname)) {
        return createFakeFd();
    }

    if (real_open64 == null) {
        real_open64 = getNextOpen("open64");
    }
    if (real_open64) |func| {
        return func(pathname.?, flags, mode);
    }
    return -1;
}
